File: palindrome_list.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        #base case
        if not head:
            return True
        
        # middle node
        slow = head
        fast = head.next
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        # headB -> 2  
        headB = slow.next
        slow.next = None
         
        # reverse
        fast = self.reverse(headB)
        
        # Check val of head and headB
        while head and fast:
            if head.val != fast.val:
                return False
            else:
                head = head.next
                fast = fast.next
        return True
        
    def reverse(self, head):
        prev = None
        curr = head
        fast = ListNode(-1)

        while curr != None:
            fast = curr.next
            curr.next = prev
            prev = curr
            curr = fast
            #fast = fast.next
            
        #curr.next = prev
        head = prev
        return head

File: Palindromelist.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 
 // find a mid
 // reverse the 2nd half of the list
 // check the palindrome with 2 ptr on both the list
 
 1 2 3
 head -> head.next
 
 TC - O(N/2) to find mid + O(N/2) for reverse + O(N/2) for palindrome check.
 SC - O(1)

 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        
        if (head == null)
        {
            return true;
        }

        ListNode mid = findMid(head);
        
        System.out.println(mid.val);
        
        ListNode nextHead = mid.next;
        mid.next = null;
        
        ListNode revereHead = reverseTheList(nextHead);
        
        ListNode first = head;
        ListNode second = revereHead;
        
        while(first != null && second != null)
        {
            if (first.val != second.val)
            {
                return false;
            }
            
            first = first.next;
            second = second.next;
        }
        
        return true;
    }
    
    /**
    findMid = 1->2->3->2->1
    
    1 2 3
    **/
    private ListNode findMid(ListNode head)
    {
        ListNode slow = head;
        ListNode fast = head.next;
        
        while(fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return slow;
    }
    
    
    // return head of the reveresed linked list
    private ListNode reverseTheList(ListNode head)
    {
        ListNode prev = null;
        ListNode current = head;
        
        while(current != null)
        {
            ListNode newNode = current.next;
            current.next = prev;
            prev = current;
            current = newNode;
        }
        
        return prev;
    }
}
